### PR TITLE
Making parse_server_name more consistent

### LINKED ISSUE #12122: `parse_and_validate_server_name(":80")` raises `IndexError` instead of `ValueError`
Currently `parse_server_name("")` raises a ValueError by accident, while `parse_server_name(":80")` returns `("", 80)`.
The latter causes `parse_and_validate_server_name` to raise an IndexError when it tries to inspect the first character of the hostname.

As part of the fix, we should make `parse_server_name` behave consistently. Either `parse_server_name("")` should return `("", None)`, or `parse_server_name(":80")` should raise a ValueError.

I'm mildly in favor of the former, since in all other cases `parse_server_name` behaves like a hostname-port splitter without regard for the validity of the hostname.

https://sentry.matrix.org/sentry/synapse-matrixorg/issues/231342/
https://sentry.matrix.org/sentry/synapse-matrixorg/issues/242910/
https://sentry.matrix.org/sentry/synapse-matrixorg/issues/197400/

### PR DESCRIPTION
**Pull request by Abdullah0sama**
Fixes #12122

### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#changelog). The entry should:
  - Be a short description of your change which makes sense to users. "Fixed a bug that prevented receiving messages from other servers." instead of "Moved X method from `EventStore` to `EventWorkerStore`.".
  - Use markdown where necessary, mostly for `code blocks`.
  - End with either a period (.) or an exclamation mark (!).
  - Start with a capital letter.
  - Feel free to credit yourself, by adding a sentence "Contributed by @github_username." or "Contributed by [Your Name]." to the end of the entry.
* [x] Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)
* [x] [Code style](https://matrix-org.github.io/synapse/latest/code_style.html) is correct
  (run the [linters](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#run-the-linters))

Signed-off-by:  Abdullah Osama <abdullahosama15@gmail.com>

### CODE REVIEW

File: changelog.d/14007.misc [hunk 1]
```diff
@@ -0,0 +1 @@
+Making parse_server_name more consistent.
```
Comment (erikjohnston) [inline changelog.d/14007.misc:1]:
```suggestion
Make `parse_server_name` consistent in handling invalid server names.
```

**Review by erikjohnston — COMMENTED** [review]
Thanks! In the `test_validate_bad_server_names` unit test can you add a check for `":80"` please?

**Review by erikjohnston — APPROVED** [review]
Thanks!

**Comment by anoadragon453** [comment]
I've merged latest develop into this PR's branch in order to make CI pass. Unit tests were failing with the following backtrace:

<details>

<summary>error log</summary>

```
Error: 
Traceback (most recent call last):
  File "/tmp/tmp.A21RKCLSkU/tests/unittest.py", line 119, in new
    return code(orig, *args, **kwargs)
  File "/tmp/tmp.A21RKCLSkU/tests/unittest.py", line 163, in setUp
    return orig()
  File "/tmp/tmp.A21RKCLSkU/tests/unittest.py", line 271, in setUp
    self.hs = self.make_homeserver(self.reactor, self.clock)
  File "/tmp/tmp.A21RKCLSkU/tests/rest/media/v1/test_url_preview.py", line 95, in make_homeserver
    hs = self.setup_test_homeserver(config=config)
  File "/tmp/tmp.A21RKCLSkU/tests/unittest.py", line 516, in setup_test_homeserver
    config_obj.parse_config_dict(config, "", "")
  File "/synapse/build/debian/matrix-synapse-py3/opt/venvs/matrix-synapse/lib/python3.10/site-packages/synapse/config/_base.py", line 794, in parse_config_dict
    self.invoke_all(
  File "/synapse/build/debian/matrix-synapse-py3/opt/venvs/matrix-synapse/lib/python3.10/site-packages/synapse/config/_base.py", line 393, in invoke_all
    res[config_class.section] = getattr(config, func_name)(*args, **kwargs)
  File "/synapse/build/debian/matrix-synapse-py3/opt/venvs/matrix-synapse/lib/python3.10/site-packages/synapse/config/repository.py", line 208, in read_config
    check_requirements("url_preview")
  File "/synapse/build/debian/matrix-synapse-py3/opt/venvs/matrix-synapse/lib/python3.10/site-packages/synapse/util/check_dependencies.py", line 181, in check_requirements
    raise ValueError(f"Synapse {VERSION} does not provide the feature '{extra}'")
builtins.ValueError: Synapse 1.69.0rc1 does not provide the feature 'url_preview'
```

</details>

**Comment by DMRobertson** [comment]
> I've merged latest develop into this PR's branch in order to make CI pass. Unit tests were failing with the following backtrace:
> error log

Xref #14079

**Pull request merged**